core/vm/instructions.go

```diff
@@ -464,28 +464,61 @@ func opGasprice(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([
 	return nil, nil
 }
 
+// opBlockhash executes the BLOCKHASH opcode pre-EIP-2935
 func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
-	num := scope.Stack.Peek()
-	num64, overflow := num.Uint64WithOverflow()
+	arg := scope.Stack.Peek()
+	arg64, overflow := arg.Uint64WithOverflow()
 	if overflow {
-		num.Clear()
+		arg.Clear()
 		return nil, nil
 	}
 	var upper, lower uint64
 	upper = interpreter.evm.Context.BlockNumber
-	if upper < 257 {
+	if upper <= params.BlockHashOldWindow {
 		lower = 0
 	} else {
-		lower = upper - 256
+		lower = upper - params.BlockHashOldWindow
 	}
-	if num64 >= lower && num64 < upper {
-		num.SetBytes(interpreter.evm.Context.GetHash(num64).Bytes())
+	if arg64 >= lower && arg64 < upper {
+		arg.SetBytes(interpreter.evm.Context.GetHash(arg64).Bytes())
 	} else {
-		num.Clear()
+		arg.Clear()
 	}
 	return nil, nil
 }
 
+// opBlockhash2935 executes for the BLOCKHASH opcode post EIP-2935 by returning the
+// corresponding hash for the blocknumber from the state, if within range.
+// The range is defined by [head - params.BlockHashHistoryServeWindow - 1, head - 1]
+// This should not be used without activating EIP-2935
+func opBlockhash2935(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]byte, error) {
+	arg := scope.Stack.Peek()
+	refundAmt := params.ColdSloadCostEIP2929
+	arg64, overflow := arg.Uint64WithOverflow()
+	if overflow {
+		arg.Clear()
+	} else {
+		var upper uint64
+		upper = interpreter.evm.Context.BlockNumber
+		if arg64 >= upper || arg64+params.BlockHashHistoryServeWindow < upper {
+			arg.Clear()
+		} else {
+			storageSlot := libcommon.BytesToHash(uint256.NewInt(arg64 % params.BlockHashHistoryServeWindow).Bytes())
+			if _, slotMod := interpreter.evm.IntraBlockState().AddSlotToAccessList(params.HistoryStorageAddress, storageSlot); !slotMod {
+				refundAmt = params.ColdSloadCostEIP2929 - params.WarmStorageReadCostEIP2929
+			}
+			interpreter.evm.intraBlockState.GetState(
+				params.HistoryStorageAddress,
+				&storageSlot,
+				arg,
+			)
+		}
+	}
+	// The gas func for this charges max (ColdSloadCostEIP2929) gas, refunding the rest here
+	interpreter.evm.intraBlockState.AddRefund(refundAmt)
```

Comment:
This doesn't look right. `opSload` doesn't refund anything.

Reply:
This one is peculiar because there is a window check on the operation. If the `arg` falls outside of the window, the state reading gas doesn't apply. Awaiting more clarity on this from EIP-2935 authors.

The refund here is to not do the condition checks on window twice - once on the dynamic gas cost function, and one on the execute function.
So, the flow here dictates that `ColdSloadCostEIP2929` must be provided before accessing the execute function.

I agree that this is a little messed up. Should I go for double check at the cost of (little) performance overhead?

Reply:
Changed now